I'd have proposed something along the lines of:
```diff
diff --git a/pkgs/development/python-modules/hdf5plugin/default.nix b/pkgs/development/python-modules/hdf5plugin/default.nix
index b8847f4e2746..3083278ef311 100644
--- a/pkgs/development/python-modules/hdf5plugin/default.nix
+++ b/pkgs/development/python-modules/hdf5plugin/default.nix
@@ -11,6 +11,10 @@
   lz4,
   zlib,
   zstd,
+  stdenv,
+  sse3Support ? stdenv.hostPlatform.sse3Support,
+  avx2Support ? stdenv.hostPlatform.avx2Support,
+  avx512Support ? stdenv.hostPlatform.avx512Support,
 }:
 
 buildPythonPackage rec {
@@ -45,6 +49,16 @@ buildPythonPackage rec {
     zstd
   ];
 
+
+  # These feature defaults can enable CPU-specific code during the build:
+  # most are detected from the build host CPU, while BMI2 defaults to enabled
+  # on Linux/Darwin. Pin them to keep the output generic and machine-independent.
+  env.HDF5PLUGIN_NATIVE = "False";
+  env.HDF5PLUGIN_SSE2 = if avx2Support then "True" else "False";
+  env.HDF5PLUGIN_SSSE3 = if sse3Support then "True" else "False";
+  env.HDF5PLUGIN_AVX2 = if avx2Support then "True" else "False";
+  env.HDF5PLUGIN_AVX512 = if avx512Support then "True" else "False";
+
   # opt-in to use use system libs instead
   env.HDF5PLUGIN_SYSTEM_LIBRARIES = lib.concatStringsSep "," [
     #"blosc" # AssertionError: 4000 not less than 4000
```

(untested, and may do weird things on cross)

--- a/pkgs/development/python-modules/hdf5plugin/default.nix
+++ b/pkgs/development/python-modules/hdf5plugin/default.nix
@@ -59,6 +59,16 @@ buildPythonPackage rec {
     "zstd"
   ];
 
+  # These feature defaults can enable CPU-specific code during the build:
+  # most are detected from the build host CPU, while BMI2 defaults to enabled
+  # on Linux/Darwin. Pin them to keep the output generic and machine-independent.
+  # https://github.com/silx-kit/hdf5plugin/blob/v6.0.0/doc/install.rst#available-options
+  env.HDF5PLUGIN_NATIVE = "False";
+  env.HDF5PLUGIN_SSE2 = "False";
+  env.HDF5PLUGIN_SSSE3 = "False";
+  env.HDF5PLUGIN_AVX2 = "False";
+  env.HDF5PLUGIN_AVX512 = "False";
+
   checkPhase = ''
     python test/test.py
   '';

